### ISSUE TITLE
BUG: conversion to M8[ns]/datetime64[ns] is slow from "M8[s]" and other time resolutions

### ISSUE THREAD

**Issue by rafiyr** [pandas-dev/pandas#35025]
- [x ] I have checked that this issue has not already been reported.

- [x ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

import pandas as pd, numpy as np, time
x = np.repeat(int(time.time()), 100_000_000)

"these are slow"
pd.Series(x.astype("M8[s]"))
pd.Series(x).astype("M8[s]")

"these are an order of magnitude faster"
pd.Series(x.astype("M8[s]").astype("M8[ns]"))
pd.Series((x*1e9).astype("M8[ns]")

#### Problem description

As long as there is no value translation involved converting/creating a datetime64[ns] array is fast as expected.  But it is an order of magnitude slower when there is a magnitude conversion.  Converting from float64 is only slightly slower than int64.

I've tested with 0.25.3, 0.25.32 and 1.0.5. And they all seem roughly consistent for those examples.  The older versions have added behavior that pd.Series((x*1e9).astype("M8[ns]"), dtype="datetime64[ns]") is also slow, but fast without the dtype arguement.

In all cases the result is the same series of datetime64[ns] values.

#### Output of ``pd.show_versions()``

python           : 3.8.3.final.0
python-bits      : 64
OS               : Linux
machine          : x86_64
processor        : x86_64
byteorder        : little
pandas           : 1.0.5
numpy            : 1.18.5
pytz             : 2020.1
dateutil         : 2.8.1

**Comment by jbrockmendel** [comment]
Why do you think this is a bug?

**Comment by rafiyr** [comment]
In the sense that it produces correct output, it is not a bug.  In the sense that performance is unexpectedly slow, I would say its an issue and I would not object to a renaming of the issue to something more appropriate.

Is there some reason to expect Series to be an order of magnitude slower than numpy for converting an array of M8[s] to M8[ns]?

**Comment by jbrockmendel** [comment]
> Is there some reason to expect Series to be an order of magnitude slower than numpy for converting an array of M8[s] to M8[ns]?

Yes.  numpy's conversion is not overflow-safe, xref https://github.com/numpy/numpy/issues/16352.  

```
arr = np.array(["2367-12-31 12:00:00"], dtype="datetime64[h]")

>>> arr.astype("datetime64[ns]")
array(['1783-06-11T12:25:26.290448384'], dtype='datetime64[ns]')

>>> pd.Series(arr)
[...]
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 2367-12-31 12:00:00
```

**Comment by rafiyr** [comment]
Ah. Is the overflow detection done with vector ops?  Just as a quick test I tried: (pd.Series(df.times.values.astype("M8[s]").astype("u8")) > x).sum()
That seems to be a pretty fast operation, and I'm clearly not trying to be efficient about it.

If there's a conflict between precision and speed, perhaps just a quick conservative test to avoid the slow full precision check?

**Comment by jbrockmendel** [comment]
> Is the overflow detection done with vector ops?

The overflow-safe conversion is done in `pandas._libs.tslibs.conversion.ensure_datetime64ns` (i.e. in cython).  If you can find ways to better optimize this, PR would be welcome.

**Comment by rafiyr** [comment]
I've played around with the code a bit.  One idea is let numpy pull out the min and max and just run the safety checks on those.  I have not tested with NATs.  Aside from that, I don't think this version looses anything else over the current code.  It would be faster do the scaling with a simple multiply (~1/3 faster on the machine I'm using today) instead of using the numpy conversion, but that would introduce more assumptions about the internal format of the datetime objects.

`
    if unit == NPY_FR_ns:
        if copy:
            arr = arr.copy()
        result = arr
    else:
        pandas_datetime_to_datetimestruct(arr.min().view("int64"), unit, &dts)
        check_dts_bounds(&dts)
        pandas_datetime_to_datetimestruct(arr.max().view("int64"), unit, &dts)
        check_dts_bounds(&dts)
        result = arr.astype("M8[ns]")
`

**Comment by jbrockmendel** [comment]
What you're describing is a three-pass process, so it can likely be improved upon.

**Comment by jorisvandenbossche** [comment]
"theoritically" speaking, this can certainly be improved (eg pyarrow can do the same cast (with bounds check) in 350 ms that takes 5s in pandas). But it would indeed need some investigation / experimentation to see what is realistic to do in pandas .

**Comment by rafiyr** [comment]
It seemed that the most expensive part of the function was the conversion to and from the time structures.  The bounds can be pre-computed or hard coded for the incoming format and test each element that way.  For my first attempt I did something like that and worked entirely in the int64 space with pre-computed min, max, and scaling factors (it was faster).  Have not tried that using the numpy comparisons for datetime64 instead.

To what extent do you want to avoid assumptions about the representation inside this function?

**Comment by jorisvandenbossche** [comment]
@rafiyr what do you mean exactly with "assumptions about the representation"? You mean the fact that is is datetime64[ns], which basically are int64's representing the ns?

**Comment by jbrockmendel** [comment]
I think something like the following is a 1-pass version of what you suggested:

```
def ensure_datetime64ns(...):

    if unit in [month, year]:
        return do_what_were_doing_now(...)
    if unit in [picos, attos]:
         return just_divide_no_need_for_overflow_check(...)
    
    factor = nanoseconds_per_unit(unit)  # i.e. 10**3 for us, 10**6 for ms, 10**9 for s,...
    upper = biggest_i8_val_that_wont_overflow(unit)
    lower = smallest_i8_val_that_wont_overflow(unit)
   
    i8result = np.empty(n, dtype="i8")
    for i in range(n):
        val = values[i]
        if val == NPY_NAT:
             i8result[i] = NPY_NAT
        else:
            if lower <= val <= upper:
                i8result[i] = val * factor
            else:
                 raise OutOfBoundsDatetime(...)
    return i8result
```

**Comment by rafiyr** [comment]
@jorisvandenbossche yup, wasn't sure if it would be considered safe enough to just produce the scaling factor and use regular multiplication.

Is there a particular reason to indent that innermost if block instead of using elif lower <= val...?

Do year and month need special handling for leap days?  Are other coarse larger units safe given leap seconds?

Also, its worth keeping the pass through for ns.

**Comment by jorisvandenbossche** [comment]
Yes, conversion between resolutions like s or ms to ns are simple scaling factors. For other resolutions like days or years, that will be different I suppose. So a more specialized / performant function can focus on the simple resolutions, and leave the slower path for other resolutions.

**Comment by jbrockmendel** [comment]
> Do year and month need special handling for leap days? Are other coarse larger units safe given leap seconds?

Year and month are not just multiplicative, so will need to keep using the existing code.  Everything higher-resolution that that should be OK with the multiplicative approach.

**Comment by rafiyr** [comment]
And that's why I asked what is safe.  In playing around it seemed like they scaled better than I'd expected given the variable length of those periods.  I'm still a bit concerned about the effect of leap seconds on minutes-weeks.

**Comment by jbrockmendel** [comment]
> I'm still a bit concerned about the effect of leap seconds on minutes-weeks.

datetime64 does not have a concept of leap seconds

**Comment by rafiyr** [comment]
Any updates on this issue?

**Comment by jbrockmendel** [comment]
@rafiyr you're welcome to make a PR

**Comment by jreback** [comment]
i thought we had a PR for this - but doesn't appear to be linked (or my memory is going)

**Comment by jbrockmendel** [comment]
xref https://github.com/numpy/numpy/issues/16352
